text/0000-close-trait.md

```diff
@@ -0,0 +1,303 @@
+- Feature Name: close-trait
+- Start Date: 2019-04-03
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add a `Close` trait to `std::io` that allows for dropping a value and producing
+a `Result`.
+
+# Motivation
+[motivation]: #motivation
+
+In many programming languages, it is possible to explicitly `close` many types
+of resources.  In Rust, we have the `Drop` trait to automatically cleanup these
+resources for us.  For various reasons, we have decided that `Drop` shall never
+fail.  Thus errors encountered in a `Drop` implementation should be ignored.
+Implementations of `Close` will allow us to explicitly handle errors that could
+occur in a `Drop` implementation.
+
+Adding the ability to manually `close` resources will allow users of the
+language to handle errors only revealed when the resource is dropped.  One
+example of this would be a race condition for shared resource access.  Another
+would be in the case of a resource that cannot be fully flushed before it is
+dropped.  It may not be possible to completely flush a resource without closing
+it.  It may also be possible that the process of closing a resource itself could
+cause an error.  In this case, there is currently no standard way to deal with
+all of these edge cases.
+
+Making `Close` into a trait will allow us to generically use this functionality
+for many types of resources.  This will also allow us to `close` wrapper types,
+such as `BufReader` or `BufWriter` when they wrap any type implementing `Close`.
+
+For specifically `File`s, we can call `sync_all` to force synchronization of the
+data to the filesystem.  But this function is no longer directly available when
+we use a `BufReader` or `BufWriter`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Currently in rust, we can manipulate files simply and expect them to be
+automatically closed:
+
+```rust
+use std::fs::File;
+use std::io::prelude::*;
+
+fn main() -> std::io::Result<()> {
+    let mut file = File::create("foo.txt")?;
+    file.write_all(b"Hello world!")?;
+    Ok(())
+}
+```
+
+Although this checks if there were errors in writing the data to the file, there
+sometimes are spurious failures that aren't caught here.  Our program is
```

Comment:
For someone like myself who might not be familiar with them, can you give some examples of what these failures might be? Are there any that `File::sync_all` wouldn’t surface?

Reply:
For specifically `File`, that is correct.  `sync_all` allows us to handle these errors.  Maybe a better example is something like a `TCPStream` where you have to send a finish packet to signal to the other side to close the connection.